include/kernel.h

```diff
@@ -0,0 +1,531 @@
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
+	reg_block_t* stack_state;
+	/** Current state of interrupt depth for this thread */
+	int interrupt_depth;
+	/** Size of the stack in bytes */
+	int stack_size;
+	/** Name of thread (for debugging purposes) */
+	char name[17];
+	/** Internal flags */
+	uint8_t flags;
+	/** Priority of the thread (0=lowest, use positive numbers only) */
+	int pri;
+	/** Intrusive link to next thread in a waiting list */
+	struct kthread_s *next;
+	/** Entry point function for the thread */
+	void (*user_entry)(void*);
+	/** Custom argument to be passed to the entry point */
+	void *user_data;
+	/** Pointer to the stack buffer */
+	void *stack;
+} kthread_t;
+
+/**
+ * @brief A kernel message box.
+ *
+ * A message box is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore.
+ *
+ * A message box can be allocated either on the heap (#mbox_new) or on
+ * the stack (#mbox_new_stack). In both cases, the size of the buffer
+ * (maximum number of mails that can be kept within the message box)
+ * must be specified.
+ *
+ * Threads can send mails using #kmbox_send or #kmbox_try_send: the former
+ * will block the thread if the message box is full, while the later will
+ * simply return a failure in that case.
+ *
+ * Threads can receive mails using #kmbox_recv or #kmbox_try_recv.
+ */
+typedef struct kmbox_s
+{
+	/** Read index of the circular buffer */
+	uint8_t r;
+	/** Write index of the circular buffer */
+	uint8_t w;
+	/** Size of the buffer (maximum number of mails) */
+	uint8_t max;
+	/** Number of events this mailbox is attached to */
+	uint8_t evts;
+	/** Linked list of threads waiting on this mailbox */
+	kthread_t *wait_list;
+	/** Circular buffer of mails */
+	void* mails[0];
+} kmbox_t;
+
+/** Maximum number of mailboxes that can be attached to a single event. */
+#define MAX_MAILBOXES_PER_EVENT 8
+
+/**
+ * @brief a kernel event.
+ *
+ * An event, represented by a #kevent_t instance, is a signal that can
+ * be broadcasted to many different threads. A thread can use
+ * #kthread_wait_event to wait for a specific event to fire. The kernel
+ * library includes a set of events bound to MI peripheral interrupts,
+ * so that it's possible for threads to wait for specific interrupts and
+ * act after them.
+ *
+ * It is also possible to attach a mailbox to an event using #mbox_attach_event.
+ * This is useful if a thread needs to wait for multiple events.
+ */
+typedef struct kevent_s
+{
+	/** Mailboxes attached to this event */
+	kmbox_t *mboxes[MAX_MAILBOXES_PER_EVENT];
+} kevent_t;
+
+
+/** @brief Event generated when a SP interrupt is triggered */
+extern kevent_t KEVENT_IRQ_SP;
```

Comment:
Would it make sense to add events for non-RCP interrupts such as the timer interrupt (IP7) or the pre-NMI interrupt (IP4)?

Reply:
Thanks, I've added pre-NMI.

Timer interrupt is more delicate; in general, timers are used by kernels for internal bookkeeping (eg: preemptive multitasking, or thread sleeping while yielding). libdragon already offers a good timer library to register callbacks, so I feel like user-level code in a thread shouldn't mess with timer interrupt. So I'd rather not expose it at this level.